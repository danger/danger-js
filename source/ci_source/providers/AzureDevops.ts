import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"

export class AzureDevops implements CISource {
  constructor(private readonly env: Env) {}
  get name(): string {
    return "AzureDevops"
  }

  get isCI(): boolean {
    return true
  }
  get isPR(): boolean {
    if (ensureEnvKeysExist(this.env, ["CI_PULL_REQUEST"]) || ensureEnvKeysExist(this.env, ["CIRCLE_PULL_REQUEST"])) {
      return true
    }

    const mustHave = ["CIRCLE_CI_API_TOKEN", "CIRCLE_PROJECT_USERNAME", "CIRCLE_PROJECT_REPONAME", "CIRCLE_BUILD_NUM"]
    return ensureEnvKeysExist(this.env, mustHave) && ensureEnvKeysAreInt(this.env, ["CIRCLE_PR_NUMBER"])
  }
  get repoSlug(): string {
    return ""
  }
  get pullRequestID(): string {
    return this.env.BITRISE_PULL_REQUEST
  }
  get commitHash(): string {
    return this.env.BITRISE_GIT_COMMIT
  }
  get ciRunURL(): string {
    return this.env.BITRISE_PULL_REQUEST
  }
  get useEventDSL(): boolean {
    return this.env.GITHUB_EVENT_NAME !== "pull_request"
  }
}
