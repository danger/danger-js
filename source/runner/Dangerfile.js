// @flow

import type { DangerResults } from "./DangerResults"
import type { DangerDSLType } from "../dsl/DangerDSL"
import type { MarkdownString } from "../dsl/Aliases"

export interface DangerContext {
/* BEGIN FLOWTYPE EXPORT */
  /**
   * Fails a build, outputting a specific reason for failing
   *
   * @param {MarkdownString} message the String to output
   */
  fail(message: MarkdownString): void;

  /**
   * Highlights low-priority issues, does not fail the build
   *
   * @param {MarkdownString} message the String to output
   */
  warn(message: MarkdownString): void;

  /**
   * Puts a message inside the Danger table
   *
   * @param {MarkdownString} message the String to output
   */
  message(message: MarkdownString): void;

  /**
   * Puts a message inside the Danger table
   *
   * @param {MarkdownString} message the String to output
   */
  markdown(message: MarkdownString): void;

  /** Typical console */
  console: any;

  /**
   * The Danger object to work with
   *
   * @type {DangerDSLType}
   */
  danger: DangerDSLType;
  /**
   * Results of a Danger run
   *
   * @type {DangerDSLType}
   */
  results: DangerResults;
/* END FLOWTYPE EXPORT */
}

/** Creates a Danger context, this provides all of the global functions
 *  which are available to the Danger eval runtime.
 *
 * @param {DangerDSLType} dsl The DSL which is turned into `danger`
 * @returns {DangerContext} a DangerContext-like API
 */
export function contextForDanger(dsl: DangerDSLType): DangerContext {
  const results: DangerResults = {
    fails: [],
    warnings: [],
    messages: [],
    markdowns: []
  }

  const fail = (message: MarkdownString) => {
    results.fails.push({ message })
  }

  const warn = (message: MarkdownString) => {
    results.warnings.push({ message })
  }

  const message = (message: MarkdownString) => {
    results.messages.push({ message })
  }

  const markdown = (message: MarkdownString) => {
    results.markdowns.push(message)
  }

  return {
    fail,
    warn,
    message,
    markdown,
    console,
    results,
    danger: dsl
  }
}
