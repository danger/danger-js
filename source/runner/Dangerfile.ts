import { DangerRuntimeContainer } from "../dsl/DangerResults"
import { DangerDSLType, PerilDSL } from "../dsl/DangerDSL"
import { MarkdownString } from "../dsl/Aliases"

/// Start of Danger DSL definition

/** A function with a callback function, which Danger wraps in a Promise */
export type CallbackableFn = (callback: (done: any) => void) => void

/**
 * Types of things which Danger will schedule for you, it's recommended that you
 * just throw in an `async () => { [...] }` function. Can also handle a function
 * that has a single 'done' arg.
 */
export type Scheduleable = Promise<any> | Promise<void> | CallbackableFn

export interface DangerContext {
  /**
   * A Dangerfile, in Peril, is evaluated as a script, and so async code does not work
   * out of the box. By using the `schedule` function you can now register a
   * section of code to evaluate across multiple tick cycles.
   *
   * `schedule` currently handles two types of arguments, either a promise or a function with a resolve arg.
   *
   * @param {Function} asyncFunction the function to run asynchronously
   */
  schedule(asyncFunction: Scheduleable): void

  /**
   * Fails a build, outputting a specific reason for failing into a HTML table.
   *
   * @param {MarkdownString} message the String to output
   * @param {string | undefined} file a file which this message should be attached to
   * @param {number | undefined} line the line which this message should be attached to
   */
  fail(message: MarkdownString, file?: string, line?: number): void

  /**
   * Highlights low-priority issues, but does not fail the build. Message
   * is shown inside a HTML table.
   *
   * @param {MarkdownString} message the String to output
   * @param {string | undefined} file a file which this message should be attached to
   * @param {number | undefined} line the line which this message should be attached to
   */
  warn(message: MarkdownString, file?: string, line?: number): void

  /**
   * Adds a message to the Danger table, the only difference between this
   * and warn is the emoji which shows in the table.
   *
   * @param {MarkdownString} message the String to output
   * @param {string | undefined} file a file which this message should be attached to
   * @param {number | undefined} line the line which this message should be attached to
   */
  message(message: MarkdownString, file?: string, line?: number): void
  /**
   * Adds a message to the Danger table, the only difference between this
   * and warn is the default emoji which shows in the table.
   * You can also specifiy a custom emoji to show in the table for each message
   * 
   * @param {MarkdownString} message the String to output
   * @param {{file?: string, line?: string, icon?: MarkdownString}} [opts]
   * @param opts.file a file which this message should be attached to
   * @param opts.line the line which this message should be attached to
   * @param opts.icon icon string or image to show in table, take care not to break table formatting
   */
  message(message: MarkdownString, opts?: { file?: string, line?: number, icon?: MarkdownString }): void

  /**
   * Adds raw markdown into the Danger comment, under the table
   *
   * @param {MarkdownString} message the String to output
   * @param {string | undefined} file a file which this message should be attached to
   * @param {number | undefined} line the line which this message should be attached to
   */
  markdown(message: MarkdownString, file?: string, line?: number): void

  /**
   * The root Danger object. This contains all of the metadata you
   * will be looking for in order to generate useful rules.
   */
  danger: DangerDSLType

  /**
   * When Peril is running your Dangerfile, the Danger DSL is
   * extended with additional options.
   */
  peril: PerilDSL

  /**
   * The current results of a Danger run, this can be useful if you
   * are wanting to introspect on whether a build has already failed.
   */
  results: DangerRuntimeContainer
}

/// End of Danger DSL definition

/** Creates a Danger context, this provides all of the global functions
 *  which are available to the Danger eval runtime.
 *
 * @param {DangerDSLType} dsl The DSL which is turned into `danger`
 * @returns {DangerContext} a DangerContext-like API
 */
export function contextForDanger(dsl: DangerDSLType): DangerContext {
  const results: DangerRuntimeContainer = {
    fails: [],
    warnings: [],
    messages: [],
    markdowns: [],
    scheduled: [],
  }

  const schedule = (fn: any) => results.scheduled && results.scheduled.push(fn)
  const fail = (message: MarkdownString, file?: string, line?: number) => results.fails.push({ message, file, line })
  const warn = (message: MarkdownString, file?: string, line?: number) => results.warnings.push({ message, file, line })

  const message = (message: MarkdownString, opts?: string | { file?: string, line?: number, icon?: MarkdownString }, lineArg?: number) => {
    let file: string | undefined
    let line: number | undefined
    let icon: MarkdownString | undefined

    if (typeof opts === 'string') {
      file = opts
      line = lineArg
    }
    else if (typeof opts === 'object') {
      ({ file, line, icon } = opts)
    }

    results.messages.push({ message, file, line, icon })
  }
  const markdown = (message: MarkdownString, file?: string, line?: number) =>
    results.markdowns.push({ message, file, line })

  return {
    schedule,
    fail,
    warn,
    message,
    markdown,
    results,
    danger: dsl,
    peril: {} as any,
  }
}
