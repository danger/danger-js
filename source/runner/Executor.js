// @flow

import { contextForDanger } from "../runner/Dangerfile"
import { DangerDSL } from "../dsl/DangerDSL"
import type { CISource } from "../ci_source/ci_source"
import { Platform } from "../platforms/platform"
import type { DangerResults } from "../runner/DangerResults"
import githubResultsTemplate from "./templates/github-issue-template"
import { runDangerfile } from "./DangerfileRunner"

// This is still badly named, maybe it really should just be runner?

export default class Executor {
  ciSource: CISource
  platform: Platform

  constructor(ciSource: CISource, platform: Platform) {
    this.ciSource = ciSource
    this.platform = platform
  }

  /**
   *  Runs all of the operations for a running just Danger
   * @returns {void} It's a promise, so a void promise
   */
  async runDanger() {
    const dsl = await this.dslForDanger()
    const context = contextForDanger(dsl)
    const results = await runDangerfile("dangerfile.js", context)
    await this.handleResults(results)
  }

  /** Sets up all the related objects for running the Dangerfile
  * @returns {void} It's a promise, so a void promise
  */
  async dslForDanger(): Promise<DangerDSL> {
    const git = await this.platform.getReviewDiff()
    const pr = await this.platform.getReviewInfo()
    return new DangerDSL(pr, git)
  }

  /**
   * Handle the messaing aspects of running a Dangerfile
   * @returns {void} It's a promise, so a void promise
   * @param {DangerResults} results a JSON representation of the end-state for a Danger run
   */
  async handleResults(results: DangerResults) {
    // Ensure process fails if there are fails
    if (results.fails.length) {
      process.exitCode = 1
    }

    // Delete the message if there's nothing to say
    const hasMessages =
      results.fails.length > 0 ||
      results.warnings.length > 0 ||
      results.messages.length > 0 ||
      results.markdowns.length > 0

    if (!hasMessages) {
      console.log("All Good.")
      await this.platform.deleteMainComment()
    } else {
      console.log("Failed")
      const comment = githubResultsTemplate(results)
      await this.platform.updateOrCreateComment(comment)
    }
  }
}
