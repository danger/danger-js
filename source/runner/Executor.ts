import { DangerContext } from "./Dangerfile"
import { CISource } from "../ci_source/ci_source"
import { Platform, Comment } from "../platforms/platform"
import {
  inlineResults,
  regularResults,
  mergeResults,
  DangerResults,
  DangerInlineResults,
  resultsIntoInlineResults,
  emptyDangerResults,
  inlineResultsIntoResults,
  sortResults,
  sortInlineResults,
  validateResults,
  isEmptyResults,
} from "../dsl/DangerResults"
import {
  template as githubResultsTemplate,
  inlineTemplate as githubResultsInlineTemplate,
  fileLineToString,
  messageForResultWithIssues as githubMessageForResultWithIssues,
} from "./templates/githubIssueTemplate"
import {
  template as bitbucketServerTemplate,
  inlineTemplate as bitbucketServerInlineTemplate,
  messageForResultWithIssues as bitbucketMessageForResultWithIssues,
} from "./templates/bitbucketServerTemplate"
import {
  template as bitbucketCloudTemplate,
  inlineTemplate as bitbucketCloudInlineTemplate,
  messageForResultWithIssues as bitbucketCloudMessageForResultWithIssues,
} from "./templates/bitbucketCloudTemplate"
import exceptionRaisedTemplate from "./templates/exceptionRaisedTemplate"

import { debug } from "../debug"
import chalk from "chalk"
import { sentence, href, compliment } from "./DangerUtils"
import { DangerRunner } from "./runners/runner"
import { GitDSL } from "../dsl/GitDSL"
import { DangerDSL } from "../dsl/DangerDSL"
import { emptyGitJSON } from "../platforms/github/GitHubGit"

export interface ExecutorOptions {
  /** Should we do a text-only run? E.g. skipping comments */
  stdoutOnly: boolean
  /** Should the output be submitted as a JSON string? */
  jsonOnly: boolean
  /** Should Danger post as much info as possible */
  verbose: boolean
  /** A unique ID to handle multiple Danger runs */
  dangerID: string
  /** Don't send the entire JSON via STDIN, instead store it in tmp, and give the paths */
  passURLForDSL: boolean
  /** Disable Checks support in GitHub */
  disableGitHubChecksSupport?: boolean
  /** Fail if danger report contains failures */
  failOnErrors?: boolean
  /** Dont add danger check to PR */
  noPublishCheck?: boolean
}
// This is still badly named, maybe it really should just be runner?

const isTests = typeof jest === "object"

interface ExitCodeContainer {
  exitCode: number
}

export class Executor {
  private readonly d = debug("executor")
  private readonly log = isTests ? () => "" : console.log
  private readonly logErr = isTests ? () => "" : console.error

  constructor(
    public readonly ciSource: CISource,
    public readonly platform: Platform,
    public readonly runner: DangerRunner,
    public readonly options: ExecutorOptions,
    public readonly process: ExitCodeContainer
  ) {}

  /**
   *  Runs all of the operations for a running just Danger
   * @param {string} file the filepath to the Dangerfile
   * @returns {Promise<DangerResults>} The results of the Danger run
   */

  async runDanger(file: string, runtime: DangerContext) {
    let results = {} as DangerResults

    // If an eval of the Dangerfile fails, we should generate a
    // message that can go back to the CI
    try {
      results = await this.runner.runDangerfileEnvironment([file], [undefined], runtime)
    } catch (error) {
      results = this.resultsForError(error)
    }

    await this.handleResults(results, runtime.danger.git)
    return results
  }

  /**
   * Sets up all the related objects for running the Dangerfile
   * @returns {void} It's a promise, so a void promise
   */
  async dslForDanger(): Promise<DangerDSL> {
    // This checks if the CI source, and the platform support running on
    // an event that's not a PR
    const useSimpleDSL = this.platform.getPlatformReviewSimpleRepresentation && this.ciSource.useEventDSL
    this.d("Using full Danger DSL:", !useSimpleDSL)

    // Can't use the API to grab git metadata
    const git = useSimpleDSL ? emptyGitJSON() : await this.platform.getPlatformGitRepresentation()

    const getDSLFunc = useSimpleDSL
      ? this.platform.getPlatformReviewSimpleRepresentation
      : this.platform.getPlatformReviewDSLRepresentation

    const platformDSL = await getDSLFunc!()

    const utils = { sentence, href }
    return new DangerDSL(platformDSL, git, utils, this.platform.name)
  }

  /**
   * Handle the message aspects of running a Dangerfile
   *
   * @param {DangerResults} results a JSON representation of the end-state for a Danger run
   */
  async handleResults(results: DangerResults, git?: GitDSL) {
    this.d("Got results back:", results)
    if (!results) {
      throw new Error(
        "Got no results back from the Dangerfile evaluation, this is likely an issue with a custom sub-process exiting early."
      )
    }
    validateResults(results)

    this.d(`Evaluator settings`, this.options)

    if (this.options.stdoutOnly || this.options.jsonOnly || (this.ciSource && this.ciSource.useEventDSL)) {
      await this.handleResultsPostingToSTDOUT(results)
    } else {
      await this.handleResultsPostingToPlatform(results, git)
    }

    if (this.options.failOnErrors && results.fails.length > 0) {
      this.process.exitCode = 1
    }
  }
  /**
   * Handle showing results inside the shell.
   *
   * @param {DangerResults} results a JSON representation of the end-state for a Danger run
   */
  async handleResultsPostingToSTDOUT(results: DangerResults) {
    const { fails, warnings, messages, markdowns } = results
    if (this.options.jsonOnly) {
      // Format for Danger Process
      const results = {
        fails,
        warnings,
        messages,
        markdowns,
      }
      process.stdout.write(JSON.stringify(results, null, 2))
    } else {
      this.d("Writing to STDOUT:", results)
      // Human-readable format

      const tick = chalk.bold.greenBright("✓")
      const cross = chalk.bold.redBright("ⅹ")
      let output = ""

      if (fails.length > 0) {
        const s = fails.length === 1 ? "" : "s"
        const are = fails.length === 1 ? "is" : "are"
        const message = chalk.underline.red("Failing the build")
        output = `Danger: ${cross} ${message}, there ${are} ${fails.length} fail${s}.`
      } else if (warnings.length > 0) {
        const message = chalk.underline("not failing the build")
        output = `Danger: ${tick} found only warnings, ${message}`
      } else if (messages.length > 0) {
        output = `Danger: ${tick} passed, found only messages.`
      } else if (!messages.length && !fails.length && !messages.length && !warnings.length) {
        output = `Danger: ${tick} passed review, received no feedback.`
      }

      const allMessages = [...fails, ...warnings, ...messages, ...markdowns].map(m => m.message)
      const oneMessage = allMessages.join("\n")
      const longMessage = oneMessage.split("\n").length > 30

      // For a short message, show the log at the top
      if (!longMessage) {
        // An empty blank line for visual spacing
        this.log(output)
      }

      const table = [
        fails.length && { name: "Failures", messages: fails.map(f => f.message) },
        warnings.length && { name: "Warnings", messages: warnings.map(w => w.message) },
        messages.length && { name: "Messages", messages: messages.map(m => m.message) },
        markdowns.length && { name: "Markdowns", messages: markdowns.map(m => m.message) },
      ].filter(r => r !== 0) as { name: string; messages: string[] }[]

      // Consider looking at getting the terminal width, and making it 60%
      // if over a particular size

      table.forEach(row => {
        this.log(`## ${chalk.bold(row.name)}`)
        this.log(row.messages.join(chalk.bold("\n-\n")))
      })

      // For a long message show the results at the bottom
      if (longMessage) {
        this.log("")
        this.log(output)
      }

      // An empty blank line for visual spacing
      this.log("")
    }
  }

  /**
   * Handle showing results inside a code review platform
   *
   * @param {DangerResults} results a JSON representation of the end-state for a Danger run
   * @param {GitDSL} git a reference to a git implementation so that inline comments find diffs to work with
   */
  async handleResultsPostingToPlatform(originalResults: DangerResults, git?: GitDSL) {
    // Allow a platform to say "I can do something special with this" - the example case for this
    // is the GitHub Checks API. It doesn't have an API that feels like commenting, so
    // it allows transforming the results after doing its work.
    let results = originalResults
    if (this.platform.platformResultsPreMapper) {
      this.d("Running platformResultsPreMapper:")
      results = await this.platform.platformResultsPreMapper(results, this.options, this.ciSource.commitHash)
      this.d("Received results from platformResultsPreMapper:", results)
    }

    const { fails, warnings, messages, markdowns } = results

    const failureCount = [...fails, ...warnings].length
    const messageCount = [...messages, ...markdowns].length

    this.d("Posting to platform:", results)

    const dangerID = this.options.dangerID
    const failed = fails.length > 0

    let issueURL = undefined

    if (failureCount + messageCount === 0) {
      this.log(`Found no issues or messages from Danger. Removing any existing messages on ${this.platform.name}.`)
      await this.platform.deleteMainComment(dangerID)
      const previousComments = await this.platform.getInlineComments(dangerID)
      for (const comment of previousComments) {
        if (comment && comment.ownedByDanger) {
          await this.deleteInlineComment(comment)
        }
      }
    } else {
      if (fails.length > 0) {
        const s = fails.length === 1 ? "" : "s"
        const are = fails.length === 1 ? "is" : "are"
        this.log(`Failing the build, there ${are} ${fails.length} fail${s}.`)
      } else if (warnings.length > 0) {
        this.log("Found only warnings, not failing the build.")
      } else if (messageCount > 0) {
        this.log("Found only messages, passing those to review.")
      }

      let mergedResults = regularResults(results)
      if (git !== undefined) {
        const previousComments = await this.platform.getInlineComments(dangerID)
        const inline = inlineResults(results)
        const inlineLeftovers = await this.sendInlineComments(inline, git, previousComments)
        mergedResults = sortResults(mergeResults(mergedResults, inlineLeftovers))
      }

      // If danger have no comments other than inline to update. Just delete previous main comment.
      if (isEmptyResults(mergedResults)) {
        this.platform.deleteMainComment(dangerID)
      } else {
        let commitID
        if (this.ciSource.commitHash !== undefined) {
          commitID = this.ciSource.commitHash
        } else if (git !== undefined) {
          commitID = git.commits[git.commits.length - 1].sha
        }
        let comment
        if (process.env["DANGER_BITBUCKETSERVER_HOST"]) {
          comment = bitbucketServerTemplate(dangerID, mergedResults, commitID)
        } else if (process.env["DANGER_BITBUCKETCLOUD_OAUTH_KEY"] || process.env["DANGER_BITBUCKETCLOUD_USERNAME"]) {
          comment = bitbucketCloudTemplate(dangerID, mergedResults, commitID)
        } else {
          comment = githubResultsTemplate(dangerID, mergedResults, commitID)
        }

        issueURL = await this.platform.updateOrCreateComment(dangerID, comment)
        this.log(`Feedback: ${issueURL}`)
      }
    }

    if (!this.options.noPublishCheck) {
      await this.updatePrStatus(!failed, issueURL, results, dangerID)
    }

    // More info, is more info.
    if (this.options.verbose) {
      await this.handleResultsPostingToSTDOUT(results)
    }
  }

  async updatePrStatus(
    passed: boolean | "pending",
    issueURL: string | undefined,
    results: DangerResults,
    dangerID: string
  ) {
    const urlForInfo = issueURL || this.ciSource.ciRunURL
    const successPosting = await this.platform.updateStatus(passed, messageForResults(results), urlForInfo, dangerID)

    if (!successPosting && this.options.verbose) {
      this.log("Could not add a commit status, the GitHub token for Danger does not have access rights.")
      this.log("If the build fails, then danger will use a failing exit code.")
    }

    if (!successPosting && !passed) {
      this.d("Failing the build due to handleResultsPostingToPlatform not successfully setting a commit status")
      process.exitCode = 1
    }
  }

  /**
   * Send inline comments
   * Returns these violations that didn't pass the validation (e.g. incorrect file/line)
   *
   * @param results Results with inline violations
   */
  sendInlineComments(results: DangerResults, git: GitDSL, previousComments: Comment[] | null): Promise<DangerResults> {
    if (!this.platform.supportsInlineComments) {
      return new Promise(resolve => resolve(results))
    }

    const inlineResults = resultsIntoInlineResults(results)
    const sortedInlineResults = sortInlineResults(inlineResults)

    let emptyResult: DangerResults = {
      messages: emptyDangerResults.messages,
      markdowns: emptyDangerResults.markdowns,
      fails: emptyDangerResults.fails,
      warnings: emptyDangerResults.warnings,
      meta: results.meta,
    }

    // For every inline result check if there is a comment already
    // if there is - update it and remove comment from deleteComments array (comments prepared for deletion)
    // if there isn't - create a new comment
    // Leftovers in deleteComments array should all be deleted afterwards
    let deleteComments = Array.isArray(previousComments) ? previousComments.filter(c => c.ownedByDanger) : []
    let commentPromises: Promise<any>[] = []
    for (let inlineResult of sortedInlineResults) {
      const index = deleteComments.findIndex(p =>
        p.body.includes(fileLineToString(inlineResult.file, inlineResult.line))
      )
      let promise: Promise<any>
      if (index != -1) {
        let previousComment = deleteComments[index]
        deleteComments.splice(index, 1)
        promise = this.updateInlineComment(inlineResult, previousComment)
      } else {
        promise = this.sendInlineComment(git, inlineResult)
      }
      commentPromises.push(promise.then(_r => emptyDangerResults).catch(_e => inlineResultsIntoResults(inlineResult)))
    }
    deleteComments.forEach(comment => {
      let promise = this.deleteInlineComment(comment)
      commentPromises.push(promise.then(_r => emptyDangerResults).catch(_e => emptyDangerResults))
    })

    return Promise.all(commentPromises).then(dangerResults => {
      return new Promise<DangerResults>(resolve => {
        resolve(dangerResults.reduce((acc, r) => mergeResults(acc, r), emptyResult))
      })
    })
  }

  async sendInlineComment(git: GitDSL, inlineResults: DangerInlineResults): Promise<any> {
    const comment = this.inlineCommentTemplate(inlineResults)
    return await this.platform.createInlineComment(git, comment, inlineResults.file, inlineResults.line)
  }

  async updateInlineComment(inlineResults: DangerInlineResults, previousComment: Comment): Promise<any> {
    const body = this.inlineCommentTemplate(inlineResults)
    // If generated body is exactly the same as current comment we don't send an API request
    if (body == previousComment.body) {
      return
    }

    return await this.platform.updateInlineComment(body, previousComment.id)
  }

  async deleteInlineComment(comment: Comment): Promise<any> {
    return await this.platform.deleteInlineComment(comment.id)
  }

  inlineCommentTemplate(inlineResults: DangerInlineResults): string {
    const results = inlineResultsIntoResults(inlineResults)

    let comment
    if (process.env["DANGER_BITBUCKETSERVER_HOST"]) {
      comment = bitbucketServerInlineTemplate(this.options.dangerID, results, inlineResults.file, inlineResults.line)
    } else if (process.env["DANGER_BITBUCKETCLOUD_OAUTH_KEY"] || process.env["DANGER_BITBUCKETCLOUD_USERNAME"]) {
      comment = bitbucketCloudInlineTemplate(this.options.dangerID, results, inlineResults.file, inlineResults.line)
    } else {
      comment = githubResultsInlineTemplate(this.options.dangerID, results, inlineResults.file, inlineResults.line)
    }
    return comment
  }

  /**
   * Takes an error (maybe a bad eval) and provides a DangerResults compatible object
   * @param error Any JS error
   */
  resultsForError(error: Error) {
    // Need a failing error, otherwise it won't fail CI.
    this.logErr(chalk.red("Danger has failed to run"))
    this.logErr(error)
    return {
      fails: [{ message: "Running your Dangerfile has Failed" }],
      warnings: [],
      messages: [],
      markdowns: [{ message: exceptionRaisedTemplate(error) }],
    }
  }
}

const messageForResults = (results: DangerResults) => {
  if (!results.fails.length && !results.warnings.length) {
    return `All green. ${compliment()}`
  } else {
    if (process.env["DANGER_BITBUCKETSERVER_HOST"]) {
      return bitbucketMessageForResultWithIssues
    } else if (process.env["DANGER_BITBUCKETCLOUD_OAUTH_KEY"] || process.env["DANGER_BITBUCKETCLOUD_USERNAME"]) {
      return bitbucketCloudMessageForResultWithIssues
    } else {
      return githubMessageForResultWithIssues
    }
  }
}
