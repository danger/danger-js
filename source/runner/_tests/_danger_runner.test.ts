import { contextForDanger } from "../Dangerfile"
import {
  createDangerfileRuntimeEnvironment,
  runDangerfileEnvironment,
  updateDangerfile,
  cleanDangerfile,
  dangerJestConfig,
} from "../DangerfileRunner"

import { FakeCI } from "../../ci_source/providers/Fake"
import { FakePlatform } from "../../platforms/FakePlatform"
import { Executor } from "../Executor"

import * as os from "os"
import * as fs from "fs"

import { resolve } from "path"
const fixtures = resolve(__dirname, "fixtures")

/**
 * Sets up an example context
 * @returns {Promise<DangerContext>} a context
 */
async function setupDangerfileContext() {
  const platform = new FakePlatform()
  const config = {
    stdoutOnly: false,
    verbose: false,
  }

  const exec = new Executor(new FakeCI({}), platform, config)

  platform.getPlatformGitRepresentation = jest.fn()
  platform.getPlatformDSLRepresentation = jest.fn()

  const dsl = await exec.dslForDanger()
  return contextForDanger(dsl)
}

describe("with fixtures", () => {
  it("handles a blank Dangerfile", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileEmpty.js"), runtime)

    expect(results).toEqual({
      fails: [],
      markdowns: [],
      messages: [],
      warnings: [],
    })
  })

  it("handles a full set of messages", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileFullMessages.js"), runtime)

    expect(results).toEqual({
      fails: [{ message: "this is a failure" }],
      markdowns: ["this is a *markdown*"],
      messages: [{ message: "this is a message" }],
      warnings: [{ message: "this is a warning" }],
    })
  })

  it("handles a failing dangerfile", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)

    try {
      await runDangerfileEnvironment(resolve(fixtures, "__DangerfileBadSyntax.js"), runtime)
      throw new Error("Do not get to this")
    } catch (e) {
      // expect(e.message === ("Do not get to this")).toBeFalsy()
      expect(e.message).toEqual("hello is not defined")
    }
  })

  it("handles relative imports correctly", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    await runDangerfileEnvironment(resolve(fixtures, "__DangerfileImportRelative.js"), runtime)
  })

  it("handles scheduled (async) code", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileScheduled.js"), runtime)
    expect(results).toEqual({
      fails: [],
      messages: [],
      markdowns: [],
      warnings: [{ message: "Asynchronous Warning" }],
    })
  })

  it("handles multiple scheduled statements and all message types", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileMultiScheduled.js"), runtime)
    expect(results).toEqual({
      fails: [{ message: "Asynchronous Failure" }],
      messages: [{ message: "Asynchronous Message" }],
      markdowns: ["Asynchronous Markdown"],
      warnings: [{ message: "Asynchronous Warning" }],
    })
  })

  // This adds > 6 seconds to the tests! Only orta should be forced into that.
  if (process.env["USER"] === "orta") {
    it("can execute async/await scheduled functions", async () => {
      // this test takes *forever* because of babel-polyfill being required
      const context = await setupDangerfileContext()
      const runtime = await createDangerfileRuntimeEnvironment(context)
      const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileAsync.js"), runtime)
      expect(results.warnings).toEqual([
        {
          message: "Async Function",
        },
        {
          message: "After Async Function",
        },
      ])
    })
  }

  it("can schedule callback-based promised", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileCallback.js"), runtime)
    expect(results.warnings).toEqual([
      {
        message: "Scheduled a callback",
      },
    ])
  })

  it("can handle TypeScript based Dangerfiles", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfileTypeScript.ts"), runtime)
    expect(results.messages).toEqual([
      {
        message: "Honey, we got Types",
      },
    ])
  })

  it("can handle a plugin (which is already used in Danger)", async () => {
    const context = await setupDangerfileContext()
    const runtime = await createDangerfileRuntimeEnvironment(context)
    const results = await runDangerfileEnvironment(resolve(fixtures, "__DangerfilePlugin.js"), runtime)

    expect(results.fails[0].message).toContain("@types dependencies were added to package.json")
  })
})

describe("cleaning Dangerfiles", () => {
  it("Supports removing the danger import", () => {
    const path = resolve(os.tmpdir(), "fake_dangerfile_1")
    fs.writeFileSync(path, "import { danger, warn, fail, message } from 'danger'")
    updateDangerfile(path)
    expect(fs.readFileSync(path).toString()).toEqual("// Removed import")
  })

  it("also handles typescript style imports", () => {
    const before = `
import { danger, warn, fail, message } from 'danger'
import { danger, warn, fail, message } from "danger"
import { danger, warn, fail, message } from "danger";
import danger from "danger"
import danger from 'danger'
import danger from 'danger';
`
    const after = `
// Removed import
// Removed import
// Removed import
// Removed import
// Removed import
// Removed import
`
    expect(cleanDangerfile(before)).toEqual(after)
  })

  it("also handles require style imports", () => {
    const before = `
const { danger, warn, fail, message } = require('danger')
var { danger, warn, fail, message } = require("danger")
let { danger, warn, fail, message } = require('danger');
`
    const after = `
// Removed require
// Removed require
// Removed require
`
    expect(cleanDangerfile(before)).toEqual(after)
  })
})

it("creates a working jest config", async () => {
  const config = await dangerJestConfig()
  // OK, this is almost perfect, but well, everyone has different paths.
  // So we'll amend the ones that should be different per developer/CI
  config.cacheDirectory = "[cache]"
  config.testPathDirs = ["[testPathDirs]"]
  config.testPathIgnorePatterns = ["[testPathIgnorePatterns]"]
  config.moduleDirectories = ["[moduleDirectories]"]

  const cwd = process.cwd()
  config.transform = config.transform.map(([files, transformer]) => {
    const trans = transformer.includes("ts-jest") ? "[ts-jest-transformer]" : transformer
    return [files, trans]
  })

  expect(config).toMatchSnapshot()
})
