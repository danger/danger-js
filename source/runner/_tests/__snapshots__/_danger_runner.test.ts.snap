// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a working jest config 1`] = `
Object {
  "cache": null,
  "cacheDirectory": "[cache]",
  "haste": Object {
    "defaultPlatform": "danger-js",
  },
  "moduleDirectories": Array [
    "[moduleDirectories]",
  ],
  "moduleFileExtensions": Array [
    "js",
    "js",
    "jsx",
    "json",
    "ts",
    "tsx",
  ],
  "moduleNameMapper": Array [],
  "name": "danger",
  "setupFiles": Array [],
  "testEnvironment": "node",
  "testPathDirs": Array [
    "[testPathDirs]",
  ],
  "testPathIgnorePatterns": Array [
    "[testPathIgnorePatterns]",
  ],
  "testRegex": "",
  "transform": Array [
    Array [
      "js$",
      "babel-jest",
    ],
    Array [
      ".(ts|tsx)",
      "[ts-jest-transformer]",
    ],
  ],
  "transformIgnorePatterns": Array [
    "/node_modules/",
  ],
}
`;
